src/wait.rs

```diff
@@ -0,0 +1,288 @@
+        loop {
+            match self.timeout {
+                Some(timeout) if start.elapsed() > timeout => {
+                    break Err(error::CmdError::WaitTimeout)
+                }
+                _ => {}
+            }
+            match condition.ready(self.client).await? {
+                Some(result) => break Ok(result),
+                None => {
+                    tokio::time::sleep(self.period).await;
+                }
+            };
+        }
+    }
+
+    /// Wait for a predicate.
+    pub async fn on_predicate<F>(self, predicate: F) -> Result<(), error::CmdError>
+    where
+        F: for<'f> FnMut(
+            &'f mut Client,
+        ) -> Pin<
+            Box<dyn Future<Output = Result<bool, error::CmdError>> + 'f + Send>,
+        >,
+    {
+        self.on(Predicate(predicate)).await
+    }
+}
+
+/// A condition to wait for.
+///
+/// This is implemented by different types directly, like [`Locator`]. But you can also use
+/// custom logic, using the [`Condition`] and [`Predicate`] newtypes.
+pub trait WaitCondition<'a, T> {
+    /// Check if the condition is ready. If it is, return `Some(...)`, otherwise return `None`.
+    fn ready(
+        &'a mut self,
+        client: &'a mut Client,
+    ) -> Pin<Box<dyn Future<Output = Result<Option<T>, error::CmdError>> + 'a + Send>>;
+}
+
+impl<'a> WaitCondition<'a, Element> for Locator<'_> {
+    fn ready(
+        &'a mut self,
+        client: &'a mut Client,
+    ) -> Pin<Box<dyn Future<Output = Result<Option<Element>, CmdError>> + 'a + Send>> {
+        let locator: webdriver::command::LocatorParameters = (*self).into();
+        Box::pin(async move {
+            match client.by(locator).await {
+                Ok(element) => Ok(Some(element)),
+                Err(error::CmdError::NoSuchElement(_)) => Ok(None),
+                Err(err) => Err(err),
+            }
+        })
+    }
+}
+
+impl<'a> WaitCondition<'a, ()> for url::Url {
+    fn ready(
+        &'a mut self,
+        client: &'a mut Client,
+    ) -> Pin<Box<dyn Future<Output = Result<Option<()>, CmdError>> + 'a + Send>> {
+        Box::pin(async move {
+            Ok(match &client.current_url().await? == self {
+                true => Some(()),
+                false => None,
+            })
+        })
+    }
+}
+
+impl<'a, F, Fut, T> WaitCondition<'a, T> for F
+where
+    F: FnMut(&'a mut Client) -> Pin<Box<Fut>>,
+    Fut: Future<Output = Result<Option<T>, error::CmdError>> + 'a + Send,
+{
+    fn ready(
+        &'a mut self,
+        client: &'a mut Client,
+    ) -> Pin<Box<dyn Future<Output = Result<Option<T>, CmdError>> + 'a + Send>> {
+        self(client)
+    }
+}
+
+/// Newtype for condition functions.
+///
+/// A condition is successful once it returns some value. It will be re-tried as long as it
+/// returns [`None`].
+///
+/// Instead of using the newtype, you can also use the [`Wait::on_condition`] method.
+#[derive(Debug)]
+pub struct Condition<F, T>(pub F)
```

Comment:
Is there a particular reason you want this type to be public? I'd just make it private. Same with `Predicate`.

Reply:
Yes, I wanted to allow this to be used a parameter to `on`. This way one can keep the pattern or just using `on` for everything: Locator, Url, …

Reply:
I think we should make it private — in practice I don't think there's that much value in supporting people passing lots of different types to `on` directly.

Reply:
I personally prefer a generic `on` method over the specialized `on_predicate` methods. If you really don't like these types, maybe we can just hide them from the documentation, and let people use them when they want.

I don't see much benefit in hiding this. I think both ways (newtype and specialized method) are valid approaches, and would simply let the user decide.

Reply:
So, there are two reasons I don't like the generic `on`:

1. It means that if, in the future, we get a more reliable way to wait for an element's appearance than just spinning with a delay, we won't be able to take advantage of it easily. For example, for locators, you could imagine using the observer API to make the browser notify us when a locator matches, rather than just retrying after a certain amount of time. Retrofitting that into `on` is tricky when it's generic, since we'd only be able to do it _specifically_ for `Locator`.
2. It means that there are multiple ways to do the same thing. That makes the API more complicated for users, and means we need to support twice as many use-cases. If we instead _just_ have dedicated methods, it's very obvious to users how they do each of the waiting operations, and they don't have to worry about the trait bounds or the `WaitCondition` trait.

So, I'm very much partial to just stripping this down to `fn on(Locator)`, `fn until(impl Fn(&mut Client) -> bool)`, and `fn until_some(impl Fn(&mut Client) -> Option<T>)`. Honestly, I would probably skip the `-> bool` version too, but could be convinced otherwise.